Document incremental annotation processing

oehme opened the pull request:
This is a rough first shot at documenting incremental annotation processors. I'd appreciate some early feedback from you @eriwen and @blindpirate

eriwen commented on subprojects/docs/src/docs/userguide/javaPlugin.adoc:
```diff
@@ -539,6 +539,55 @@ The incremental compilation at a high level:
+=== Incremental annotation processing
+
+Starting with Gradle 4.7, the incremental compiler also supports incremental annotation processing.
+Annotation processors need to opt-in to this feature, otherwise they will result in full recompilation.
```
> will result in

I think "will trigger" is better here

oehme replied:
Done

eriwen commented on subprojects/docs/src/docs/userguide/javaPlugin.adoc:
```diff
@@ -539,6 +539,55 @@ The incremental compilation at a high level:
+
+==== Making an annotation processor incremental
+
+Gradle supports incremental compilation for two common categories of annotation processors.
```
> two common categories of annotation processors.

I think it would be good to write `two common categories of annotation processors: "Isolated" and "Aggregating".`

oehme replied:
Done

eriwen commented on subprojects/docs/src/docs/userguide/javaPlugin.adoc:
```diff
@@ -539,6 +539,55 @@ The incremental compilation at a high level:
+==== Making an annotation processor incremental
+
+Gradle supports incremental compilation for two common categories of annotation processors.
+Processors that don't fit these categories will result in full recompilation as before.
```
> Processors that don't fit these categories

Is there a specific category or example we can name here? 

Also: "as before" is superfluous here.

oehme replied:
Well, any processors that do any of the things listed as limitations below. E.g. those that have untracked side effects, those that are nondeterministic (yes, that's a thing), those that read/write resources. I'll try to fit this in here.

eriwen commented on subprojects/docs/src/docs/userguide/javaPlugin.adoc:
```diff
@@ -539,6 +539,55 @@ The incremental compilation at a high level:
+As a processor author, please consult the information below to decide which category your processor falls in.
+Then add a file called
+
+`META-INF/gradle/incremental.annotation.processors`
```
Any particular reason this gets a paragraph of its own?

eriwen commented on subprojects/docs/src/docs/userguide/javaPlugin.adoc:
```diff
@@ -539,6 +539,55 @@ The incremental compilation at a high level:
+
+and add one line for each of your incremental processors and their category, separated by a comma.
+
+    org.example.MyProcessor,ISOLATED
```
:o: we might optionally consider one of those thingies with a full example title like https://github.com/gradle/gradle/blame/master/subprojects/docs/src/docs/userguide/workingWithFiles.adoc#L52-L56 just to call this out more.

oehme replied:
Done

eriwen commented on subprojects/docs/src/docs/userguide/javaPlugin.adoc:
```diff
@@ -539,6 +539,55 @@ The incremental compilation at a high level:
+Isolated processors have the following limitations:
+
+* Can't read resources
+* Can't write resources (this may change in the future)
```
Let's omit "this may change in the future" from everything we write, please

oehme replied:
Got it

eriwen commented on subprojects/docs/src/docs/userguide/javaPlugin.adoc:
```diff
@@ -539,6 +539,55 @@ The incremental compilation at a high level:
+
+===== Isolated annotation processors
+These look at each annotated element in isolation, creating generated files or validation messages for it.
+For instance a `RepositoryProcessor` could create a repository implementation for each interface annotated with `@Repository`
```
> repository implementation

I think it would be clearer to be very specific: "create a `RepositoryProcessorImpl`"

Even better if we can have _some_ visual here like a little diagram of input => outputs

I think it's important that folks really, really understand this, lest they apply the wrong category for their custom annotation processors.

oehme replied:
I think I'll just write a full code sample for both kinds of processors.

eriwen commented on subprojects/docs/src/docs/userguide/javaPlugin.adoc:
```diff
@@ -539,6 +539,55 @@ The incremental compilation at a high level:
+
+===== Aggregating annotation processors
+These aggregate several source files into one ore more output files or validation message.
+For example, a `ServiceRegistryProcessor` might create a single `ServiceRegistry` with one method for each type annotated with `@Service`
```
This is excellent.

:x: We should end with a period, though.

eriwen replied:
Extra super bonus points for a diagram here.

eriwen commented on subprojects/docs/src/docs/userguide/javaPlugin.adoc:
```diff
@@ -539,6 +539,55 @@ The incremental compilation at a high level:
+* Can't have any side effects except for using the `Filer` and `Messager` APIs
+
+Gradle will always recompile all files that contain annotations which the aggregating processor registered itself for.
+Gradle will also always recompile all files generated by an aggregating processor.
```
We probably could join this sentence with the previous one.

oehme commented:
@eriwen I've addressed all your feedback except for splitting it into a different chapter. Do you think we should split out just this topic or everything about the incremental compiler or everything about compiler optimisations?

geftimov commented on subprojects/docs/src/samples/java/incrementalAnnotationProcessing/processor/src/main/java/EntityProcessor.java:
```diff
@@ -0,0 +1,51 @@
+        return true;
+    }
+
+    private void createRepsitory(TypeElement entity) {
```
A typo. Should be `createRepository()`

eriwen commented:
@oehme I think we should extract anything that is not specific to Java. IIRC our incremental compilation is specific to Java, but also IIRC this incremental annotation processing is useful for Kotlin, Groovy, etc. 

I trust your judgement whether to extract it and where. In general, I fully trust you to ensure the feature is successful, including that there's enough docs and users are able to find them.

blindpirate commented on subprojects/docs/src/docs/userguide/javaPlugin.adoc:
```diff
@@ -538,7 +538,69 @@ The incremental compilation at a high level:
+=== Incremental annotation processing
 
+Starting with Gradle 4.7, the incremental compiler also supports incremental annotation processing.
+Annotation processors need to opt-in to this feature, otherwise they will trigger a full recompilation.
```
I think `opt-in` is a noun and `opt in` is verb.

blindpirate commented:
@oehme All looks good, except one thing: do we have any tests for `incrementalAnnotationProcessing` sample? I can't find them.

oehme commented:
Thanks for all the feedback! 

@blindpirate I added a test for both isolating and aggregating processors.

nniesen commented:
At least for me, there are a lot of apparent gaps in the documentation. I have thus far been unsuccessful in updating our annotation processor to work with Gradle 4.6. The biggest issue I've had is figuring out where all the pieces go for creating and using my own annotation processor. A comprehensive example would be extremely helpful. Something showing the project(s) and source set(s) to produce the custom annotation, the custom annotation compiler/processor, and then how to apply them my project's source to generate source code that will get compiled. Most of the examples are geared toward Android/dagger or are not updated for Gradle 4.6. One of the biggest points of confusion for me with the classpath separation is where to put the "resources/META-INF" stuff. Does it go in the processor's source set or in the consuming project's source set?

Our current setup was created using 2.10 and everything where the everything was thrown in the classpath and the jar dependency was added to downstream testCompiles. I'd like to split this out properly for Gradle 4.6+. We have an annotation processor that we use in JUnit tests to generate fluent Hamcrest matcher builders for the given class(es) bean properties.

For example; this generates 2 matcher builders that are subsequently used in this tests source to construct hamcrest matchers for assertThat and thrown.expect:
```
@MatcherBuilderSupport({ SampleClass.class, IllegalStateException.class })
public class SampleClassTest {
    @Test
    public void testFoo() throws Exception {
        [...]
        assertThat(actual, SampleClassMatcherBuilder.SampleClass().name("Abe").age(60).height(greaterThan(65)).matchesCompletely());
        [...]
        this.thrown.expect(IllegalStateExceptionMatcherBuilder.IllegalStateException().message(containsString("Some Detail")).matchesPartially());
        [...]
    }
}
```

If I'm understanding correctly, I should have something like the following projects to separate everything out as recommended:

## matcherbuilder-annotations
main/java/src/com/matcherbuilder/annotations/MatcherBuilderSupport.java
produces: matcherbuilder-annotations.jar

## matcherbuilder-compiler
dependencies { implementation "matcherbuilder-annotations" }
main/java/src/com/matcherbuilder/compiler/MatcherBuilderProcessor.java
produces: matcherbuilder-compiler.jar

## some-java-project
test/java/src/com/test/SampleClassTest.java (from above example)
**build.gradle**
```
configurations {
    processors
}
dependencies {
    processors 'matcherbuilder-compiler'

    testCompileOnly 'matcherbuilder-annotations'
}
sourceSets {
    test {
        java {
            // additional source folder for the generated MatcherBuilder<ClassName>.java files
            srcDir "build/generated/src/test/java"
        }
    }
}
compileTestJava {
    options.annotationProcessorPath = configurations.processors
    options.annotationProcessorGeneratedSourcesDirectory = "build/generated/src/test/java"
}
```

## Which projects and source sets get these files?
- `resources/META-INF/services/javax.annotation.processing.Processor`
- `resources/META-INF/gradle/incremental.annotation.processors`

So far I've just been trying to wire the above with just multiple source sets instead of projects. I've tried many combinations and I either get: nothing generated, the processors on the compile classpath error, or the compile classpath error that generated but said it couldn't use the generated files in the same source (can't remember how I did that one).

Thanks, Nathan.

oehme commented:
Hey Nathan,

I think your questions should be answered by the example that I added in this PR. It contains exactly what you're asking for: A multi-project build with an annotation, processor and user of that processor. Just download the [full distribution](http://services.gradle.org/distributions-snapshots/gradle-4.7-20180323002511+0000-all.zip) for the 4.7 nightly and look into the `samples/java/incrementalAnnotationProcessing` directory.

Wiring this with sourceSets is also possible, but more complex. Better use projects to separate these concerns :)

nniesen commented:
@oehme Thanks, that provided the clarity I needed on where the resources should be located. I also discovered I missed the built-in "annotationProcessor" dependency configuration. My code is now split out into separate projects and I can get the processor to run for the main source set. Although, I actually want the processor to run on various test source sets, not main, and I'm having issues with that; which I haven't fully tracked down.

"Newer" 3+? versions of Gradle don't seem to clean out previously generated code (see last comment on https://github.com/immutables/immutables/issues/134) resulting in the "IllegalStateException: endPosTable already set" error below. Ironically, this is the error that had me trying to restructure our annotation processor to be Gradle 5 compliant.

```
> Task :compileJava FAILED
MatcherBuilderSupportAnnotationProcessor process processingOver: false
Note: Processing @MatcherBuilderSupport on com.foo.test.NestedCauseMatcher
Note: Processing @MatcherBuilderSupport value=java.lang.Throwable
An exception has occurred in the compiler (1.8.0_162). Please file a bug against the Java compiler via the Java bug reporting page (http://bugreport.java.com) after checking the Bug Database (http://bugs.java.com) for duplicates. Include your program and the following diagnostic in your report. Thank you.
java.lang.IllegalStateException: endPosTable already set
        at com.sun.tools.javac.util.DiagnosticSource.setEndPosTable(DiagnosticSource.java:136)
        at com.sun.tools.javac.util.Log.setEndPosTable(Log.java:350)
        at com.sun.tools.javac.main.JavaCompiler.parse(JavaCompiler.java:667)
        at com.sun.tools.javac.main.JavaCompiler.parseFiles(JavaCompiler.java:950)
        at com.sun.tools.javac.processing.JavacProcessingEnvironment$Round.<init>(JavacProcessingEnvironment.java:892)
        at com.sun.tools.javac.processing.JavacProcessingEnvironment$Round.next(JavacProcessingEnvironment.java:921)
        at com.sun.tools.javac.processing.JavacProcessingEnvironment.doProcessing(JavacProcessingEnvironment.java:1187)
        at com.sun.tools.javac.main.JavaCompiler.processAnnotations(JavaCompiler.java:1170)
        at com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:856)
        at com.sun.tools.javac.main.Main.compile(Main.java:523)
        at com.sun.tools.javac.api.JavacTaskImpl.doCall(JavacTaskImpl.java:129)
        at com.sun.tools.javac.api.JavacTaskImpl.call(JavacTaskImpl.java:138)
        at org.gradle.api.internal.tasks.compile.JdkJavaCompiler.execute(JdkJavaCompiler.java:50)
        at org.gradle.api.internal.tasks.compile.JdkJavaCompiler.execute(JdkJavaCompiler.java:36)
        at org.gradle.api.internal.tasks.compile.NormalizingJavaCompiler.delegateAndHandleErrors(NormalizingJavaCompiler.java:98)
        at org.gradle.api.internal.tasks.compile.NormalizingJavaCompiler.execute(NormalizingJavaCompiler.java:51)
        at org.gradle.api.internal.tasks.compile.NormalizingJavaCompiler.execute(NormalizingJavaCompiler.java:37)
        at org.gradle.api.internal.tasks.compile.CleaningJavaCompilerSupport.execute(CleaningJavaCompilerSupport.java:35)
        at org.gradle.api.internal.tasks.compile.CleaningJavaCompilerSupport.execute(CleaningJavaCompilerSupport.java:25)
        at org.gradle.api.tasks.compile.JavaCompile.performCompilation(JavaCompile.java:156)
        at org.gradle.api.tasks.compile.JavaCompile.compile(JavaCompile.java:141)
        at org.gradle.api.tasks.compile.JavaCompile.compile(JavaCompile.java:114)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:73)
        at org.gradle.api.internal.project.taskfactory.IncrementalTaskAction.doExecute(IncrementalTaskAction.java:50)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:39)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:26)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$1.run(ExecuteActionsTaskExecuter.java:124)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:317)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:309)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:185)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:97)
        at org.gradle.internal.operations.DelegatingBuildOperationExecutor.run(DelegatingBuildOperationExecutor.java:31)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeAction(ExecuteActionsTaskExecuter.java:113)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:95)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:73)
        at org.gradle.api.internal.tasks.execution.OutputDirectoryCreatingTaskExecuter.execute(OutputDirectoryCreatingTaskExecuter.java:51)
        at org.gradle.api.internal.tasks.execution.SkipUpToDateTaskExecuter.execute(SkipUpToDateTaskExecuter.java:59)
        at org.gradle.api.internal.tasks.execution.ResolveTaskOutputCachingStateExecuter.execute(ResolveTaskOutputCachingStateExecuter.java:54)
        at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.execute(ValidatingTaskExecuter.java:59)
        at org.gradle.api.internal.tasks.execution.SkipEmptySourceFilesTaskExecuter.execute(SkipEmptySourceFilesTaskExecuter.java:101)
        at org.gradle.api.internal.tasks.execution.FinalizeInputFilePropertiesTaskExecuter.execute(FinalizeInputFilePropertiesTaskExecuter.java:44)
        at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:91)
        at org.gradle.api.internal.tasks.execution.ResolveTaskArtifactStateTaskExecuter.execute(ResolveTaskArtifactStateTaskExecuter.java:62)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:59)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:54)
        at org.gradle.api.internal.tasks.execution.ExecuteAtMostOnceTaskExecuter.execute(ExecuteAtMostOnceTaskExecuter.java:43)
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:34)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker$1.run(DefaultTaskGraphExecuter.java:256)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:317)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:309)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:185)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:97)
        at org.gradle.internal.operations.DelegatingBuildOperationExecutor.run(DelegatingBuildOperationExecutor.java:31)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:249)
        at org.gradle.execution.taskgraph.DefaultTaskGraphExecuter$EventFiringTaskWorker.execute(DefaultTaskGraphExecuter.java:238)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:104)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:98)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.execute(DefaultTaskExecutionPlan.java:663)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionPlan.executeWithTask(DefaultTaskExecutionPlan.java:596)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.run(DefaultTaskPlanExecutor.java:98)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
        at java.lang.Thread.run(Thread.java:748)
```

With the doFirst "fix" to delete the generated source directory, I think I'm down to some errors, that didn't used to be an issue about library dependencies but I'm still looking into these:
```
[...]SampleClass.java:16: error: For @MatcherBuilderSupport to work the class org.junit.runner.Description must be compiled and in the classpath. Check the classpath and make sure the class is not currently being compiled.
```
for sample code:
```
@MatcherBuilderSupport({ Description.class })
class SampleClass<...> {
...
```

oehme commented:
> Although, I actually want the processor to run on various test source sets, not main, and I'm having issues with that; which I haven't fully tracked down.

There shouldn't be any difference except for adding the processors to the corresponding configurations, e.g. `testAnnotationProcessor` etc.

> "Newer" 3+? versions of Gradle don't seem to clean out previously generated code

Gradle does remove the generated sources. If this doesn't happen for some reason, can you please provide a reproducible example project?

The last error doesn't really make sense to me, again an example project would help a lot.

nniesen commented:
> Gradle does remove the generated sources. If this doesn't happen for some reason, can you please provide a reproducible example project?

It breaks when you configure a different output directory. We don't wan't the generated source files in the jar so we do this for all java source in an afterEvaluate. 

Modify the `samples/java/incrementalAnnotationProcessing/user/build.gradle` to configure a different output directory:
```
compileJava {
    options.incremental = true
    
    String relativePath = "build/generated/src/main/java"
    sourceSets.main.java { srcDir relativePath }
    File generatedSourceDir = project.file("build/generated/src/main/java")
    println "Creating and configuring generated source directory $generatedSourceDir"
    project.mkdir(generatedSourceDir)
    options.annotationProcessorGeneratedSourcesDirectory = generatedSourceDir
    //outputs.dir(generatedSourceDir) // we also set this but seems unrelated
    
//    doFirst {
//        // force a clean of generatedSourceDir to prevent java.lang.IllegalStateException: endPosTable already set
//        project.delete(generatedSourceDir)
//        project.mkdir(generatedSourceDir)
//    }
}
```
`gradlew user:cJ` // works first time
`gradlew user:cJ rerun-tasks` // compiler exception

oehme commented:
I see, it doesn't work for full builds because of [this line](https://github.com/gradle/gradle/blob/master/subprojects/platform-base/src/main/java/org/gradle/language/base/internal/tasks/SimpleStaleClassCleaner.java#L32). I'll fix this for 4.7.

nniesen commented:
I also figured out what was going on with this error:
> [...]SampleClass.java:16: error: For @MatcherBuilderSupport to work the class org.junit.runner.Description must be compiled and in the classpath. Check the classpath and make sure the class is not currently being compiled.

Our annotation processor reflectively generates Hamcrest Matchers for a given class. Previously, the annotation processor had access to everything on our compile classpath but now it's separate. I'm not sure how common this is for most annotation processors but for my situation I need the processor to have access to every library on the compile classpath. Your simplistic sample just writes a bunch of string data so it doesn't encounter these issues. Anyway, it seems I can work around the issue by adding my configurations compile classpath to the annotation processors classpath as shown below but it seems like a common issue that the tooling should address.
```
dependencies {
    testAnnotationProcessor configurations.testCompile
}
```

nniesen commented:
> I see, it doesn't work for full builds because of this line. I'll fix this for 4.8.
Awsome, now I'm glad I posted; even if not in the "correct" forum.

Sorry about the interleaved posts, I've been working on the previous throughout the day.

nniesen commented:
> I see, it doesn't work for full builds because of this line. I'll fix this for 4.8.
Is `build/classes/<sourceSet>/java` really a sensible default for `options.annotationProcessorGeneratedSourcesDirectory` to write .java source files?

It seems like Gradle should automatically handle putting these source files into a `build/generated/src/<sourceSet>/java` directory to avoid polluting the compiled classes directory with source code.

Similarly, it would seem that the sourceSets compile classpath should automatically be included in the annotation processor's classpath so that the annotation processor can effectively operate against the code it is attempting to both compile and generate code for.

oehme commented:
> It would also seem a reasonable default that the sourceSets compile classpath should automatically be included in the annotation processor's classpath so that the annotation processor can effectively operate against the code it is attempting to both compile and generate code for.

A processor shouldn't try to load classes using its own classloader. It should use the RoundEnvironment to find types.

oehme commented:
Regarding the generated sources - I'm actually confused by that exception, because the javac documentation clearly states that it will overwrite existing files and doesn't require them to be removed. So this might be a compiler bug. Nevertheless we should clean them up to avoid stale files of course.

nniesen commented:
> A processor shouldn't try to load classes using its own classloader. It should use the Types and Elements APIs to find them on the compile classpath instead.

Thanks, that's good to know especially since I'm new to this and didn't write it. I'll use my hack to get started and refactor once everything is working again.

>> It seems like Gradle should automatically handle putting these source files into a build/generated/src//java directory to avoid polluting the compiled classes directory with source code.

> I agree and it should be added as a source folder in the IDEs as well. Would you mind opening a feature request for this?

See: https://github.com/gradle/gradle/issues/4956

oehme commented:
Note that your workaround will disable compile avoidance, making your builds quite a bit slower. Compile avoidance ensures that classpath changes are only considered if they change the API of a class and implementation changes are ignored. But that isn't applied to the processor path, since on the processor path, implementation details matter. So fixing your processor will have a nice performance boost on incremental builds as well.
